src/session.rs

```diff
@@ -0,0 +1,375 @@
+        self.prog.pop_stmt();
+        self.submit(InterpreterRequest::PopProgStmt);
+
+        self.var_count -= 1;
+
+        self.recompute_var_visibility();
+    }
+
+    /// Edits a program statement at the index.
+    pub fn set_prog_stmt_at(&mut self, index: usize, stmt: Stmt) {
+        // This is because the current session could want to report
+        // errors and we would like to show them somewhere
+        assert!(
+            !self.interpreter_busy(),
+            "Can't submit a request while the interpreter is already interpreting",
+        );
+
+        self.prog.set_stmt_at(index, stmt.clone());
+
+        self.submit(InterpreterRequest::SetProgStmtAt(index, stmt));
+        self.recompute_var_visibility();
+    }
+
+    /// Returns the statements currently contained in the current pipeline's
+    /// program.
+    pub fn stmts(&self) -> &[Stmt] {
+        self.prog.stmts()
+    }
+
+    /// Returns the definitions of all known functions.
+    pub fn function_table(&self) -> &BTreeMap<FuncIdent, Box<dyn Func>> {
+        &self.function_table
+    }
+
+    /// Returns the next free variable identifier for the current
+    /// program definition.
+    pub fn next_free_var_ident(&self) -> VarIdent {
+        VarIdent(self.var_count)
+    }
+
+    /// Returns human readable variable name for a variable identifier
+    /// or `None` if the variable identifier does not exist in the
+    /// current program.
+    pub fn var_name_for_ident(&self, var_ident: VarIdent) -> Option<&str> {
+        // FIXME: @Optimization Don't iterate all the time
+        self.stmts()
+            .iter()
+            .find_map(|stmt| match stmt {
+                Stmt::VarDecl(var_decl) => {
+                    if var_decl.ident() == var_ident {
+                        Some(var_decl)
+                    } else {
+                        None
+                    }
+                }
+            })
+            .map(|var_decl| {
+                self.function_table[&var_decl.init_expr().ident()]
+                    .info()
+                    .return_value_name
+            })
+    }
+
+    /// Returns all visible variable identifiers from a position
+    /// (index) in the program.
+    pub fn var_visibility_at_stmt(&self, index: usize) -> &[VarIdent] {
+        &self.var_visibility[0..index]
+    }
+
+    /// Returns whether the interpreter is currently running. Program
+    /// modifications and running the interpreter (again) are
+    /// disallowed in this state.
+    pub fn interpreter_busy(&self) -> bool {
+        !self.interpreter_requests_in_flight.is_empty()
+    }
+
+    /// Starts the interpreter on the current program.
+    pub fn interpret(&mut self) {
+        // This is because the current session could want to report
+        // errors and we would like to show them somewhere
+        assert!(
+            !self.interpreter_busy(),
+            "Can't submit a request while the interpreter is already interpreting",
+        );
```

Comment:
Good question. I don't think it *really* needs to be, but I was worried it could lead to some unpredictable behavior if the UI didn't uphold this, so it was mostly a debugging tool to make sure I didn't make a mistake in the UI code. I am not against removing, if that's what you suggest.

Reply:
It can stay as a reminder to fix this later on if you don't wanna do anything about it right now.